### PR TITLE
Don't create spurious dirs on init

### PR DESCRIPTION
**Pull request by emilypi**
This gets rid of that pesky `exe` dir being created when we specify libraries interactively: 

```
λ P temp → λ git ignore-app-dir-for-libraries → ../../dist-newstyle/build/x86_64-osx/ghc-8.10.3/cabal-install-3.5.0.0/x/cabal/noopt/build/cabal/cabal init 
...
What does the package build:
   1) Executable
   2) Library
   3) Library and Executable
Your choice? 2
...
Generating LICENSE...
Generating CHANGELOG.md...
Generating src/MyLib.hs...
Generating test/MyLibTest.hs...
Generating temp.cabal...

Warning: no synopsis given. You should edit the .cabal file and add one.
You may want to edit the .cabal file and add a Description field.
λ P temp → λ git ignore-app-dir-for-libraries → ls
CHANGELOG.md LICENSE      src          temp.cabal   test
```

and for Libraries + Executables, 

```
λ P temp → λ git ignore-app-dir-for-libraries → ../../dist-newstyle/build/x86_64-osx/ghc-8.10.3/cabal-install-3.5.0.0/x/cabal/noopt/build/cabal/cabal init 
Should I generate a simple project with sensible defaults? [default: y] n
What does the package build:
   1) Executable
   2) Library
   3) Library and Executable
Your choice? 3
...
Generating LICENSE...
Generating CHANGELOG.md...
Generating src/MyLib.hs...
Generating exe/Main.hs...
Generating test/MyLibTest.hs...
Generating temp.cabal...

Warning: no synopsis given. You should edit the .cabal file and add one.
You may want to edit the .cabal file and add a Description field.
λ P temp → λ git ignore-app-dir-for-libraries → ls
CHANGELOG.md LICENSE      exe          src          temp.cabal   test
```

and for Executables, 

```
λ P temp → λ git ignore-app-dir-for-libraries* → ../../dist-newstyle/build/x86_64-osx/ghc-8.10.3/cabal-install-3.5.0.0/x/cabal/noopt/build/cabal/cabal init
Should I generate a simple project with sensible defaults? [default: y] n
What does the package build:
   1) Executable
   2) Library
   3) Library and Executable
Your choice? 1
...
Generating LICENSE...
Generating CHANGELOG.md...
Generating exe/Main.hs...
Generating temp.cabal...

Warning: no synopsis given. You should edit the .cabal file and add one.
You may want to edit the .cabal file and add a Description field.
λ P temp → λ git ignore-app-dir-for-libraries* → ls
CHANGELOG.md LICENSE      exe          temp.cabal
```

---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [?] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.

### CODE REVIEW

**Comment by emilypi** [comment]
Question about the changelog: master seemes to only be as recent as 3.2. Do you want to just cherrypick this, or should i add an entry?

**Comment by fgaz** [comment]
> master seemes to only be as recent as 3.2


@emilypi what do you mean?

**Comment by emilypi** [comment]
@fgaz the changelog for cabal-install ends with 3.2. It doesn't seem to have been maintained. https://github.com/haskell/cabal/blob/master/cabal-install/changelog

**Comment by fgaz** [comment]
I think that's because 3.2 is  still the last release and 3.4 is still at the rc stage. @phadej?

File: cabal-install/tests/UnitTests/Distribution/Client/Init.hs [hunk 1]
```diff
@@ -104,7 +134,9 @@ exeFlags = baseFlags {
   -- Create an executable only, with main living in app/Main.hs.
     packageType = Flag Executable
   , mainIs = Flag "Main.hs"
+  , sourceDirs = Just []
```
Comment (emilypi) [inline cabal-install/tests/UnitTests/Distribution/Client/Init.hs:137]: `sourceDirs = Nothing` in `baseFlags` was actually the wrong state for this to be in. Throughout the course of `init` it actually gets turned into `Just []` for `Executable`s. So, to test against this, we set this as our golden data.

File: cabal-install/TESTING.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# Testing `cabal-install`
```
Comment (fgaz) [inline cabal-install/TESTING.md:1]: Should this be merged with CONTRIBUTING.md#running-tests?

Comment (emilypi) [reply cabal-install/TESTING.md:1]: Your folks' preference. I'd think to look in the project first for project-level testing details, but that's just me.

Comment (fgaz) [reply cabal-install/TESTING.md:1]: I agree it makes sense to have detailed info here. I'd put a link to this file in the testing section of CONTRIBUTING though, and mention cabal-testsuite here too

File: cabal-install/TESTING.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+Local testing
+=======
+
+In order to effectively test the `cabal-install` library, the `cabal-install.cabal` file must be modified
+to build the targets in the `/test` directory. The current recommended way to set this up is to
+use the [makefile](../Makefile) supplied in `Cabal` project parent directory, issuing the following command:
+
+
+```
+> make cabal-install-dev
```
Comment (fgaz) [inline cabal-install/TESTING.md:12]: The target is ambiguous

File: cabal-install/TESTING.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+```
+
+This command will copy the dev `.cabal` generated by a project build into the `cabal-install.cabal`, and set your git index to ignore
+any changes to that file. Any subsequent changes to the `.cabal` should unset and reset the git index to make sure you don't end up committing it.
+From there, tests may be built with `cabal test` as usual. To choose a particular test so you don't end up running the whole thing, you can issue
+`tasty`-style pattern expressions like the following:
+
+```
+> cabal run tests -- -p /my pattern/
```
Comment (fgaz) [inline cabal-install/TESTING.md:21]: The target is ambiguous

**Comment by fgaz** [comment]
Could you split the TESTING.md stuff into a separate commit?

File: cabal-install/TESTING.md (hunk 2 above)
Comment (emilypi) [reply cabal-install/TESTING.md:1]: Done

File: cabal-install/TESTING.md (hunk 4 above)
Comment (emilypi) [reply cabal-install/TESTING.md]: Made it something valid and runnable.

**Comment by emilypi** [comment]
@fgaz split up for you.

**Comment by emilypi** [comment]
I found a few more bugs while going through the code and wrote up a proposal fix for #7265. In case that doesn't get done, by 3.4 release, I think this would still be a good quality of life improvement for this particular function.

File: cabal-install/cabal-install.cabal.dev [hunk 5]
```diff
@@ -351,7 +351,7 @@ Test-Suite unit-tests
     UnitTests.Distribution.Client.Get
     UnitTests.Distribution.Client.Glob
     UnitTests.Distribution.Client.GZipUtils
-    UnitTests.Distribution.Client.Init.FileCreators
+    UnitTests.Distribution.Client.Init
```
Comment (fgaz) [inline cabal-install/cabal-install.cabal.dev:354]: This is a generated file. This modification should be done in the .zinza file, then the .dev file should be generated from that (there are instructions at the beginning of this file)

**Review by fgaz — CHANGES_REQUESTED** [review]
Also please replace the current subject lines of the commit messages (commit x of n) with short summaries

File: cabal-install/cabal-install.cabal.dev (hunk 5 above)
Comment (emilypi) [reply cabal-install/cabal-install.cabal.dev:354]: Ah perfect. Okay

**Comment by emilypi** [comment]
@fgaz can I get a 👍 ?

**Review by fgaz — APPROVED** [review]
:+1: :+1: 

Remember to add the changelog entry before merging!

**Comment by emilypi** [comment]
Thanks @fgaz! Added, and merging.

**Comment by ulysses4ever** [comment]
Does this merge mean that #6772 is fixed (in `master`) now? Should it be closed?

**Comment by emilypi** [comment]
Yes, good call @ulysses4ever

**Pull request merged**